azure-iot-device/azure/iot/device/common/error_map_paho.py

```diff
@@ -0,0 +1,110 @@
+# --------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+from . import errors
+from . import errors_paho
+import paho.mqtt.client as mqtt
+
+# mapping of Paho conack rc codes to Error object classes
+paho_conack_rc_to_error = {
+    mqtt.CONNACK_REFUSED_PROTOCOL_VERSION: errors_paho.PahoConnectionRefusedProtocolVersionError,
+    mqtt.CONNACK_REFUSED_IDENTIFIER_REJECTED: errors_paho.PahoConnectionRefusedIdentifierRejectedError,
+    mqtt.CONNACK_REFUSED_SERVER_UNAVAILABLE: errors_paho.PahoConnectionRefusedServerUnavailableError,
+    mqtt.CONNACK_REFUSED_BAD_USERNAME_PASSWORD: errors_paho.PahoConnectionRefusedBadUsernamePasswordError,
+    mqtt.CONNACK_REFUSED_NOT_AUTHORIZED: errors_paho.PahoConnectionRefusedNotAuthorizedError,
+}
+
+# mapping of Paho rc codes to Error object classes
+paho_rc_to_error = {
+    mqtt.MQTT_ERR_NOMEM: errors_paho.PahoNoMemoryError,
+    mqtt.MQTT_ERR_PROTOCOL: errors_paho.PahoProtocolError,
+    mqtt.MQTT_ERR_INVAL: errors_paho.PahoInvalidArgsError,
+    mqtt.MQTT_ERR_NO_CONN: errors_paho.PahoNoConnectionError,
+    mqtt.MQTT_ERR_CONN_REFUSED: errors_paho.PahoConnectionRefusedError,
+    mqtt.MQTT_ERR_NOT_FOUND: errors_paho.PahoNotFoundError,
+    mqtt.MQTT_ERR_CONN_LOST: errors_paho.PahoConnectionLostError,
+    mqtt.MQTT_ERR_TLS: errors_paho.PahoTLSError,
+    mqtt.MQTT_ERR_PAYLOAD_SIZE: errors_paho.PahoPayloadSizeError,
+    mqtt.MQTT_ERR_NOT_SUPPORTED: errors_paho.PahoNotSupportedError,
+    mqtt.MQTT_ERR_AUTH: errors_paho.PahoAuthError,
+    mqtt.MQTT_ERR_ACL_DENIED: errors_paho.PahoACLDeniedError,
+    mqtt.MQTT_ERR_UNKNOWN: errors_paho.PahoUnknownError,
+    mqtt.MQTT_ERR_ERRNO: errors_paho.PahoUnknownError,
+    mqtt.MQTT_ERR_QUEUE_SIZE: errors_paho.PahoQueueSizeError,
+}
+
+# mapping of Paho-specific errors to more generic errors
+paho_error_to_generic_error = {
+    errors_paho.PahoConnectionRefusedProtocolVersionError: errors.TransportError,
+    errors_paho.PahoConnectionRefusedIdentifierRejectedError: errors.TransportError,
```

Comment:
I am thinking more in lines of `ConnectionFailedError` for these 2 as well

Reply:
They are connection failures, but it makes no sense to retry, so I separated them out.  I think I'll make something like "ConnectionFailedNoRetry" to be explicit about it.

---
In reply to: [302699678](https://github.com/Azure/azure-iot-sdk-python-preview/pull/136#discussion_r302699678) [](ancestors = 302699678)